src/flint/fmpz_factor.jl

```diff
@@ -18,30 +18,45 @@ function _factor(a::ZZRingElem)
 end
 
 # Just to give a helpful error message if someone tries to factor a boolean
-function factor(b::Bool)
-  throw(DomainError("Cannot factorize a boolean"));
+function factor(::Bool)
+  throw(DomainError("Cannot factorize a boolean"))
 end
 
-# This function handles machine integer types (up to 64 bits)
-function factor(a::T) where T <: Union{Int, UInt}
+function factor(::Type{Bool}, ::Int)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+function factor(::Type{Bool}, ::ZZRingElem)
+  throw(DomainError("Cannot have a factorization into booleans"))
+end
+
+
+# The main dispatch function: flintify gives either Int or ZZRingElem,
+# then use Julia's dispatcher to select the worker function.
+function factor(a::T)  where T <: Integer
+  iszero(a) && throw(ArgumentError("Argument must be non-zero"))
+  return factor(typeof(a), Nemo.flintify(a))
+end
+
+# Two worker functions: one for Int, one for ZZRingElem
+function factor(::Type{T}, a::Int)  where T <: Int
```

Comment:
```suggestion
function factor(::Type{T}, a::Int)  where T <: Integer
```
otherwise this type parameter `T` does nothing

Reply:
OK, for symmetry I can allow `T` to be an integer type larger than `Int` -- it is "pointless", but that's no reason to forbid it.  I am undecided whether these two worker functions should be publicly visible -- I cannot think of any genuine use case, but maybe my imagination is too weak.

Reply:
this change is also needed for something like `factor(UInt8, 32)` aka each `Integer` except `Int`

Reply:
Allowing `T` to be any `Integer`, was kind of the point behind suggesting this variant of `factor` (as an optimization)

Reply:
There should be test cases for that -- but not `factor(UInt8, 32)` (which indeed I'd consider internal only) but rather for e.g. `factor(UInt8(32))` etc., i.e. factoring elements of a bunch of different `Integer` subtypes and verifying the return values have the right type and values.

Reply:
To sum up the discussion John and me just had: to handle `UInt` we should split this, something like this:
```suggestion
function factor(::Type{T}, a::Int) where T <: Integer
  abs_a = reinterpret(UInt, a < 0 ? -a : a) # abs(a) gives the negative value a if a == typemin(Int)
  fac = factor(T, a_abs)
  if a < 0
    fac.unit = T(-1)
  end
  return fac
end

function factor(::Type{T}, a::UInt) where T <: Integer
  # ... here go on as before but of course `u = T(1)` and we don't need `abs_a` anymore
```